mitiq/folding_cirq.py

```diff
@@ -1,76 +1,197 @@
-    sub_indices = list(range(reduced_depth))
-    return fold_gates(circuit, sub_indices)
+    for (moment_index, moment) in enumerate(circuit):
+        for gate_index in range(len(moment)):
+            # TODO: It could be expensive to call fold_gate...(...) which makes a deepcopy of the circuit each call.
+            #  ==> Possible fix: Have fold_gate_at_index_in_moment(...) modify the circuit in place.
+            folded = fold_gate_at_index_in_moment(folded, moment_index + moment_shift, gate_index)
+            moment_shift += 2
+            num_folded += 1
+            if num_folded == num_to_fold:
+                return folded
 
 
-def fold_gates(circuit: Circuit, sub_indices: List[int]) -> Circuit:
-    """Applies the map G -> G G^dag G to a subset of gates of the input circuit
-    determined by sub_indices."""
+def fold_gates_at_random(circuit: Circuit, stretch: float, **kwargs) -> Circuit:
+    """Returns a folded circuit by applying the map G -> G G^dag G to a random subset of gates in the input circuit.
 
-    out = Circuit()
-    for j, gate in enumerate(circuit):
-        out += gate
-        if j in sub_indices:
-            out += inverse(gate)
-            out += gate
-    return out
+    The folded circuit has a number of gates approximately equal to stretch * n where n is the number of gates in
+    the input circuit.
 
+    Args:
+        circuit: Circuit to fold.
+        stretch: Factor to stretch the circuit by. Any real number in the interval [1, 3].
 
-def local_folding(circuit: Circuit, stretch: float, stretcher: Callable=left_stretcher) -> Circuit:
-    """Applies the map G -> G G^dag G to a subset of gates of the input circuit.
-    Returns a circuit of depth approximately equal to stretch*len(circuit).
-    The stretch factor can be any real number >= 1."""
+    Note:
+        Folding a single gate adds two gates to the circuit, hence the maximum stretch factor is 3.
+    """
+    folded = deepcopy(circuit)
 
-    if not (stretch >= 1):
-        raise ValueError("The stretch factor must be a real number >= 1.")
+    if np.isclose(stretch, 1.0, atol=1e-2):
+        return folded
 
-    if stretch <= 3:
-        return stretcher(circuit, stretch)
-    else:
-        # recursive iterations for stretch > 3
-        _ = local_folding(circuit, 3, stretcher)
-        return local_folding(_, stretch / 3, stretcher)
+    if not 1 < stretch <= 3:
+        raise ValueError("The stretch factor must be a real number between 1 and 3.")
 
+    if "seed" in kwargs.keys():
+        np.random.seed(kwargs.get("seed"))
 
-# ==================================================
-# Circuit level folding
-# ==================================================
+    ngates = len(list(folded.all_operations()))
+    num_to_fold = int(ngates * (stretch - 1.0) / 2.0)
 
+    for _ in range(num_to_fold):
+        # TODO: This allows for gates to be folded more than once, and folded gates to be folded.
```

Comment:
My intuition is that gates should be allowed to be sampled more than once, but that one should not sample from the new folds. Do you agree @andreamari I think this change is made doing:
```
        moment_index = np.random.choice(len(circuit))
        gate_index = np.random.choice(len(circuit[moment_index]))
```

Reply:
I have the same intuition, not 100% sure though. I'll wait to hear Andrea's and other's thoughts, it's not too hard to change either way.

Reply:
My intuition is that it is better to distribute the folding gates as much as possible along the circuit, so I think it is better not to sample the same gate of the original circuit (unless `stretch >3`) AND not to sample previous folds.  
This is not so easy especially because 2 new moments are created for each gate folding.
So, before solving this issue I think it is more important to decide if this proliferation of moments is good or bad. Isn't more natural to fold a gate within its moment? In other words isn't more natural that the number of input and output moments is equal for local folding?